Adds RSA-SHA1 method for the OAuth Authentication (issue #8889)

TheFRedFox opened the pull request:
That PR is providing a possible solution for the issue https://github.com/cakephp/cakephp/issues/8889.

It is my first time working with OAuth authentication, so maybe there are still some wrong things.
E.g. I don't know if the consumerSecret and tokenSecret are not required at all for RSA-SHA1 signing.
I know that the token is mandatory for 2-legged Authentication.

Method were added/changed:
- Oauth::authentication()
- Oauth::_rsaSha1()

Tests were added:
- OauthTest::testRsaSigning()
- OauthTest::testRsaSigningWithPassphraseString()
- OauthTest::testRsaSigningWithPassphraseFile()

And some files were added for the tests:
- key.pem (the private key without passphrase)
- key_with_passphrase.pem (the private key with passphrase)
- several key_passphrase with different line endings to support the tests for all operating systems

TheFRedFox committed:
adds RSA-SHA1 functionality to the OAuth class

also makes consumerSecret, token and tokenSecret not mandatory for RSA-SHA1 signing

Squashed commits:

[d48aa1e] adds RSA-SHA1 functionality to the OAuth class

also makes consumerSecret, token and tokenSecret not mandatory for RSA-SHA1 signing

Squashed commit:

[64585dc] better handling of the passphrase, now really get just the first/current line of the file

[f220600] adds RSA-SHA1 functionality to the OAuth class

TheFRedFox committed:
adds tests for the rsa-sha1 method and files needed for the tests

Tests:
 * testRsaSigning
 * testRsaSigningWithPassphraseString
 * testRsaSigningWithPassphraseFile

Files:
 * key.pem (private key without passphrase)
 * key_with_passphrase.pem (private key with passphrase)
 * key_passphrase_(crlf) (passphrase file with cr, crlf and lf line ending

TheFRedFox commented:
Ha, forgot to check phpcs on the OauthTest class. :innocent: Will fix that tomorrow.

Secondly, it seems there is still some failure with the different line endings of the passphrase files. :confused: I will also check that tomorrow.

markstory commented on src/Network/Http/Auth/Oauth.php:
```diff
@@ -135,6 +152,64 @@ protected function _hmacSha1($request, $credentials)
+     *
+     * @param \Cake\Network\Http\Request $request The request object.
+     * @param array $credentials Authentication credentials.
+     * @return string
```
PHPCS will want an `@throws` doc block tag.

TheFRedFox replied:
Actually PHPCS didn't complain about it. But sure, I will add it.

markstory commented on src/Network/Http/Auth/Oauth.php:
```diff
@@ -135,6 +152,64 @@ protected function _hmacSha1($request, $credentials)
+            throw new \RuntimeException('RSA-SHA1 signature method requires the OpenSSL extension.');
+        }
+
+        $nonce = isset($credentials['nonce']) ? $credentials['nonce'] : uniqid();
```
I think you want `uniqid(true)`.

TheFRedFox replied:
I copied the nonce default creation from the `_hmacSha1` method. Should it be `uniqid(true)` in this method too? The `_plaintext` method also just has `uniqid()`.
Btw. we can also use `Security::randomBytes()` like in #8890. :wink:

markstory replied:
:+1: `Security::randomBytes()` is probably better.

markstory commented on tests/TestCase/Network/Http/Auth/OauthTest.php:
```diff
@@ -213,4 +213,120 @@ public function testHmacSigning()
+                       'file' => 'vacaction.jpg',
+                       'size' => 'original'
+                   ]);
+        $private_key_path = TEST_APP . DS . 'config' . DS . 'key.pem';
```
Should be `$privateKeyPath`.

markstory commented on tests/TestCase/Network/Http/Auth/OauthTest.php:
```diff
@@ -213,4 +213,120 @@ public function testHmacSigning()
+                       'file' => 'vacaction.jpg',
+                       'size' => 'original'
+                   ]);
+        $private_key_path = TEST_APP . DS . 'config' . DS . 'key_with_passphrase.pem';
```
Should be `$privateKeyPath`.

markstory commented on tests/TestCase/Network/Http/Auth/OauthTest.php:
```diff
@@ -213,4 +213,120 @@ public function testHmacSigning()
+        if(PHP_EOL == "\n") $passphrase_path = TEST_APP . DS . 'config' . DS . 'key_passphrase_lf';
+        else if(PHP_EOL == "\r\n") $passphrase_path = TEST_APP . DS . 'config' . DS . 'key_passphrase_crlf';
+        else if(PHP_EOL == "\r") $passphrase_path = TEST_APP . DS . 'config' . DS . 'key_passphrase_cr';
+        else { $this->markTestSkipped('The file for the key passphrase could not be loaded as PHP_EOL could not be recognized.'); return; }
```
Can we just use `\n`. Doing platform dependent line endings is going to make deploying windows built code/files to a nix server 'work in dev, but not prod'. If we always use consistent line endings the pem file will either work, or not consistently.

TheFRedFox replied:
Therefore my idea was to provide three different files with three different line endings, so the tests can always run. I guess it failed, because while pushing the files my git client automatically convert the file. But it would also be ok with me to make the test skip if PHP_EOL is anything but `\n`.

TheFRedFox replied:
Ah, I just remember that git clients also may convert the line endings while pulling/cloning. I will change it then, that just `\n` is supported, otherwise the test will be skipped.

TheFRedFox committed:
changes all uniqid() to Security::randomBytes(16)

TheFRedFox committed:
changes option names to camelCase and adds a throw mention in doc block

TheFRedFox committed:
makes privateKeyFile mandatory for RSA-SHA1 method

TheFRedFox committed:
changes the test so that just \n line ending is supported

otherwise the test will be skipped
also some other reformats

TheFRedFox commented:
I changed the `uniqid()` to `Security::randomBytes(16)`. Is length of 16 enough, or should it be more?

I removed the other passphrase files again, and I really hope the line endings were pushed correctly this time. ^^

markstory commented:
> Is length of 16 enough

That should be plenty. Its a nonce :)

markstory commented on src/Network/Http/Auth/Oauth.php:
```diff
@@ -54,10 +49,33 @@ public function authentication(Request $request, array $credentials)
 
+            case 'RSA-SHA1':
+                if (!isset($credentials['privateKeyFile'])) {
+                    return;
```
We should throw an exception here. The developer has made a bad mistake.

TheFRedFox replied:
But I think this would be inconsistent to the other checks. It is also a bad mistake if the developer forgot the consumerKey. I would then suggest to consistently throw an exception whenever a mandatory credential key is missed.

markstory replied:
Good point. We can approach that problem separately.

TheFRedFox replied:
Sry, but you confused me a bit now. :smile: You mean, we can do it in another PR? So we can finish this PR now? ^^

markstory replied:
Yes, a separate pull request at some point in the future.

markstory commented on tests/TestCase/Network/Http/Auth/OauthTest.php:
```diff
@@ -213,4 +213,125 @@ public function testHmacSigning()
+
+        if (PHP_EOL != "\n") {
+            $this->markTestSkipped('Just the line ending ("\n") is supported. You can run the test again e.g. on a linux system.');
+            return;
```
You don't need a return here. `markTestSkipped()` throws an exception.

TheFRedFox replied:
Well, I actually knew that .. I like the `return` there because the developer or rather I can clearly see that it will definitely end there (and also it is also easier for the IDE).
But nevertheless, if you guys never use `return` there I will remove it again. :wink:

markstory replied:
We also have a `skipIf()` helper, that takes a boolean, and conditionally skips the test.

```php
$this->skipIf(PHP_EOL != "\n", 'Just the line ending ("\n") is supported. You can run the test again e.g. on a linux system.');
```
could be used.

TheFRedFox committed:
uses TestCase::skipIf() to reduce code and bring it to top

TheFRedFox commented on src/Network/Http/Auth/Oauth.php:
```diff
@@ -135,6 +153,66 @@ protected function _hmacSha1($request, $credentials)
+            rewind($resource);
+            $credentials['privateKeyPassphrase'] = $passphrase;
+        }
+        $privateKey = openssl_pkey_get_private(file_get_contents($credentials['privateKeyFile']), $credentials['privateKeyPassphrase']);
```
Do we want to change `privateKeyFile` to `privateKey` and to support `string` and `file` representations of the private key? One more opportunity for the developer to pass the private key.
And it would be consistent to the privateKeyPassphrase option (`string` and `file` representation).

markstory replied:
:+1: to consistency. Having one option that reads resources/strings means we can avoid the awkward scenario where both are defined.

TheFRedFox committed:
changes the way to pass the private key (string or resource now)

In that way it is consistent to the passed passphrase option, which is a string or resource too.

TheFRedFox committed:
adds three more tests to check the private keys as strings

TheFRedFox commented:
Alright. Made it consistent to the passphrase parameter now and added three more test to check this functionality.
I guess I am finished now. ^^
